src/game/sound_samples.c

```diff
@@ -10,14 +10,92 @@
+{
+    // TODO: Select audio specification from a menu
+    // TODO: Use a seperate callback function
+    SDL_AudioSpec destination_spec = { // stereo float32 44100Hz
+        .format = AUDIO_F32,
+        .channels = 2,
+        .freq = 44100
+    }, actual;
+    // TODO: a return value by SDL_GetNumAudioDevices that is <= 0 may not indicate an error
+    if (SDL_GetNumAudioDevices(0) <= 0) {
+        log_fail("No audio in 2019 LULW\n");
+        return -1;
+    }
+    
+    sound_samples->audio_buf_array = PUSH_LT(sound_samples->lt, nth_calloc(sound_samples->samples_count, sizeof(uint8_t*)), free);
+    if (sound_samples->audio_buf_array == NULL) {
+        log_fail("Failed to allocate memory for audio buffer pointer array\n");
+        return -1;
+    }
+    sound_samples->audio_buf_size_array = PUSH_LT(sound_samples->lt, nth_calloc(sound_samples->samples_count, sizeof(uint32_t)), free);
+    if (sound_samples->audio_buf_size_array == NULL) {
+        log_fail("Failed to allocate memory for audio buffer size array\n");
+        return -1;
+    }
+    for (size_t i = 0; i < sound_samples->samples_count; ++i) {
+        uint8_t *wav_buf; uint32_t wav_buf_len; SDL_AudioSpec wav_spec;
+
+        log_info("Loading audio file %s...\n", sample_files[i]);
+        if (SDL_LoadWAV(sample_files[i], &wav_spec, &wav_buf, &wav_buf_len) == NULL) {
+            log_fail("Load WAV file failed: %s\n", SDL_GetError());
+            return -1;
+        }
+        PUSH_LT(sound_samples->lt, wav_buf, SDL_FreeWAV);
+        SDL_AudioCVT cvt;
+        int result = SDL_BuildAudioCVT(&cvt, wav_spec.format, (uint8_t)wav_spec.channels, (int)wav_spec.freq, 
+                          destination_spec.format, (uint8_t)destination_spec.channels, (int)destination_spec.freq);
+        if (result < 0) {
+            log_fail("SDL_BuildAudioCVT failed: %s\n", SDL_GetError());
+            return -1;
+        } else if (result == 0) { // no need to do conversion
+            sound_samples->audio_buf_array[i] = wav_buf;
+            sound_samples->audio_buf_size_array[i] = wav_buf_len;
+        } else {
+            cvt.len = (int)wav_buf_len;
+            cvt.buf = PUSH_LT(sound_samples->lt, malloc((size_t)(cvt.len * cvt.len_mult)), free);
+            if (cvt.buf == NULL) {
+                log_fail("Allocating buffer for conversion failed\n");
+                return -1;
+            }
+            memcpy(cvt.buf, wav_buf, (size_t)cvt.len);
+            SDL_FreeWAV(wav_buf);
+            RELEASE_LT(sound_samples->lt, wav_buf);
+            if (SDL_ConvertAudio(&cvt) < 0) {
+                log_fail("SDL_ConvertAudio failed: %s\n", SDL_GetError());
+                return -1;
+            }
+            sound_samples->audio_buf_array[i] = cvt.buf;
+            sound_samples->audio_buf_size_array[i] = (uint32_t)cvt.len_cvt;
+        }
+    }
+    
+    sound_samples->dev = SDL_OpenAudioDevice(NULL, 0, &destination_spec, &actual, 0);
```

Comment:
It would be good to push the dev to LT.

Reply:
Unfortunately when I do this I got this compiler error:
```
In file included from /home/nashesvobodnoye/github/nothing-fork/src/game/sound_samples.c:9:0:
/home/nashesvobodnoye/github/nothing-fork/src/game/sound_samples.c: In function ‘init_buffer_and_device’:
/home/nashesvobodnoye/github/nothing-fork/src/system/lt.h:49:17: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
     lt_push(lt, (void*)res, (Dtor)dtor)
                 ^
/home/nashesvobodnoye/github/nothing-fork/src/game/sound_samples.c:87:26: note: in expansion of macro ‘PUSH_LT’
     sound_samples->dev = PUSH_LT(sound_samples->lt, SDL_OpenAudioDevice(NULL, 0, &destination_spec, NULL, 0), SDL_CloseAudioDevice);
                          ^~~~~~~
/home/nashesvobodnoye/github/nothing-fork/src/game/sound_samples.c:87:24: error: assignment makes integer from pointer without a cast [-Werror=int-conversion]
     sound_samples->dev = PUSH_LT(sound_samples->lt, SDL_OpenAudioDevice(NULL, 0, &destination_spec, NULL, 0), SDL_CloseAudioDevice);
                        ^
cc1: all warnings being treated as errors
```
It seems that void* and SDL_AudioDeviceID has two different sizes. Alternatively I can use the legacy SDL_OpenAudio and SDL_CloseAudio functions (probably what I would end up to do).

Reply:
@The-Renaissance got it! Don't worry about it then. I have a couple of ideas on how it can be done. Will demonstrate it on tomorrow's stream.

Reply:
Nah, whatever. Let's keep it like that.